proxy/src/main/java/com/velocitypowered/proxy/config/VelocityConfiguration.java

```diff
@@ -435,16 +435,34 @@ public static VelocityConfiguration read(Path path) throws IOException {
     CommentedFileConfig defaultConfig = CommentedFileConfig.of(tmpFile, TomlFormat.instance());
     defaultConfig.load();
 
-    // Retrieve the forwarding secret. First, from environment variable, then from config.
+    // Inform the user that forwarding-secret parameter has been deprecated.
+    if (config.contains("forwarding-secret")) {
+      logger.warn("The \"forwarding-secret\" configuration parameter has been deprecated due to security"
+          + " concerns. Please remove it and use the \"forwarding-secret-file\" parameter or"
+          + " \"VELOCITY_FORWARDING_SECRET\""
+          + " environment variable instead.");
+    }
+
+    // Retrieve the forwarding secret.
+    // First, from environment variable, then from config (deprecated), then from file.
     byte[] forwardingSecret;
     String forwardingSecretString = System.getenv()
         .getOrDefault("VELOCITY_FORWARDING_SECRET", config.get("forwarding-secret"));
-    if (forwardingSecretString == null || forwardingSecretString.isEmpty()) {
-      forwardingSecretString = generateRandomString(12);
-      config.set("forwarding-secret", forwardingSecretString);
+
+    // Ensure that the "forwarding-secret-file" configuration parameter and auto-created file are present.
+    String forwardingSecretFileName = config.get("forwarding-secret-file");
+    if (forwardingSecretFileName == null || forwardingSecretFileName.isEmpty()) {
```

Comment:
If the config contains the `forwarding-secret` option, we shouldn't create the secret file: its contents will not be used and the user could be confused because the config and the file contents are different. Instead, we can move this down to line 462.

Reply:
> If the config contains the `forwarding-secret` option, we shouldn't create the secret file: its contents will not be used and the user could be confused because the config and the file contents are different. Instead, we can move this down to line 462.

That would make sense as well. Though we need to iron out the file creation details (mentioned above) before that logic is moved around.